design/036-package_id.md

```diff
@@ -0,0 +1,107 @@
+# Proposal: Conan 2.0 requirements effect in PackageID
+
+| **Status**        |                                                   |
+|:------------------|:--------------------------------------------------|
+| **RFC #**         | [036](https://github.com/conan-io/tribe/pull/36)  |
+| **Submitted**     | 2022-06-17                                        |
+| **Tribe votes**   |                                                   |
+
+
+## Summary
+
+This proposal is about the effect that the dependencies have over the current package ``package_id``, but not about other factors, such as the package settings, options, configuration or other type of binary compatibility resulting from other factors other than dependencies.
+
+Conan 2.0 proposes a new default ``package_id`` that will be dynamic, based on two main cases:
```

Comment:
I'm thinking about an alternative wording to emphasize on how dependency affects binary, something like:

... based on whether dependency affects binaries in the package or merely a dependency which needs to be carried to further stage.

The examples of the former are:
1. Header-only libraries with inline or template code used in the package.
2. Static library linked into a shared library or an application.

In these cases dependency directly affects resulting binary and potentially any change in the dependency may produce a different binary. Therefore in this case "full_mode" is a reasonable default.

However if dependency is just carried to the next stage, for example:
1. An application or a shared library linked to another shared library. Then binary is affected only after dynamic linker loaded everything in runtime.
2. A static library depending on another static library. Then binary is affected only when linker produces a shared library or an application.
3.  A header-only library depending on another library. Then binary is affected only when compiler instantiates the code.

In this case it is possible that binary from upstream package may not affect binary of direct dependent downstream package. And by default Conan will use a "minor_mode" policy, which means requiring a rebuild if the major or minor of the dependency changes, but not requiring it if the patch of the dependency changes.